File: dotnet/Clipper.Domain/Entities/Clipping.cs
using System;
using Clipper.Domain.Base;
using Clipper.Domain;

namespace Clipper.Domain.Entities
{
    public class Clipping : Entity
    {
        public Clipping(Book book, ClippingType type, int page, int locationStart, int locationEnd, DateTime date, string text)
        {
            DomainValidator
                .Create()
                .When(type == ClippingType.None, Resources.ClippingShouldBeHighlightOrBookmark)
                .When((type == ClippingType.Highlight) && text.IsNullOrEmpty(), Resources.HighlightShouldHaveANonNullOrEmptyText)
                .When(Book.IsNull(), Resources.BookShouldntBeEmpty)
                .ThrowExceptionIfAnyErrors();

            Book = book;
            Type = type;
            Page = page;
            LocationStart = locationStart;
            LocationEnd = locationEnd;
            Date = date;
            Text = text;
        }

        public Book Book { get; private set; }
        public ClippingType Type { get; private set; }
        public int Page { get; private set; }
        public int LocationStart { get; private set; }
        public int LocationEnd { get; private set; }
        public DateTime Date { get; private set; }
        public string Text { get; private set; }
    }
}

File: frontend/src/models/Clipping.ts
import Book from './Book';
import ClippingType from './ClippingType';

export default class Clipping {
  id: number;
  text: string;
  book: Book;
  type: ClippingType;
  page: number;
  locationStart: number;
  locationEnd: number;
  date: Date;

  constructor(
    text: string,
    book: Book,
    type: ClippingType,
    id?: number,
    page?: number,
    locationStart?: number,
    locationEnd?: number,
    date?: Date
  ) {
    this.id = id || 0;
    this.text = text;
    this.book = book;
    this.type = type;
    this.page = page || 0;
    this.locationStart = locationStart || 0;
    this.locationEnd = locationEnd || 0;
    this.date = date || new Date();
  }
}
